runtime/JavaScript/package.json

```diff
@@ -51,13 +51,13 @@
   "exports": {
     ".": {
       "node": {
-        "types": "src/index.node.d.ts",
+        "types": "./src/antlr4/index.d.ts",
         "import": "./dist/antlr4.node.mjs",
         "require": "./dist/antlr4.node.cjs",
```

Comment:
Using `types` like this isn't exactly correct - but many packages do it and usually you can "get away with it". There are subtle issues with it though! When checking the package with https://arethetypeswrong.github.io (a site created by the TS team member who is responsible for module resolution) then you might see "Masquerading as CJS" problem being flagged on the package (not now though, but rather with this fix in place - now you have different problems reported there). 

I believe though that "Masquerading as CJS" problem should be fixed separately - or not at all. My preference is to fix it but it might complicate your setup so I'm not sure if your team would actually consider fixing it as something that you want to do.

Reply:
If you ask me, JS itself masquerades common sense, and TS only partially fixes the problem
:-)
That said we definitely want small dedicated PRs, this is critical to understanding side effects.
Re this one, why do we even need a "types" entry in the "exports" section ? Is the top-level "types" not sufficient for tsc and ts tools (I haven't explored that tbh) ?

Reply:
There are 2 main ideas behind `package.json#exports`:
- list what you allow to be consumed from the package
- allow "selective" resolution based on conditions

So in node and other runtimes, once you add `package.json#exports` you essentially "hide" everything that is not listed. This basically works like that in all runtimes capable of resolving `package.json#exports` (it's a somewhat new addition, first added in node 12). So even the old `package.json#main` field has no effect in a package with `exports` (`exports` just take priority over it)

So it only makes sense that you have to also list types here (if you don't rely on the "sibling lookup" that I've mentioned before in the other comment). However, TS decided **not to enable** `package.json#exports` by default. You need to opt into it. So as long as you are not using the new module resolution options in TS, then your package seems to be working fine - the runtime looks into `package.json#exports`, resolves files according to the manifest there and that's fine. At the same time, `package.json#exports` are ignored by default in TS so it looks into `package.json#types` (using the "classic" node module resolution algorithm).

But once you opt-into the `package.json#exports` resolutions then `package.json#types` become **ignored** in packages with `package.json#exports`. I think that this is the right thing to do (as per the first idea behind `package.json#exports` that i listed. The bizarre part though is that TS decided to not include `package.json#exports` support by default. This cases a lot of confusion because most people stick to `moduleResolution: node` (they might not even know about the newer options), they might add `package.json#exports` **withoout** the `types` condition there and they might even recheck that with their local configuration and can only assume that things work OK since it works OK in their project. I can only presume that TS team decided to do this because otherwise many packages would just break/wouldnt be able to resolve types after consumers would upgrade their TS version (people already shipped `package.json#exports` in packages before TS implemented support for that). I'm not sure if that decision was worth it... but hey, here we are 🤷‍♂️

Reply:
Finally an explanation I can understand :-)